### DISCUSSION TITLE
Passing custom fields as properties into layouts

### DISCUSSION THREAD

**Discussion by kbrookes** [joomlatools/joomlatools-pages#545] [General]
I'm in the process of building out a set of partials that I re-use, but I think I could be doing it in a more elegant way.

For example I'm building a set of hero header sections in different styles, selectable by the user, with parameters set by the user, using custom fields. For example:

- Hero? (yes/no to add a hero or not)
- Type (selects from different partial layouts)
- Title override (overrides the name of the article to present a more engaging title).
- Hero copy (additional copy)
- Text colour
- Background overlay & opacity

Right now, this means that I have a block on the Page that's assigning the values, and another in the Layout, and then finally utilising them in the Partial.

### Page
```php
$articleID = '1';
$fields = article($articleID)->fields;

$heroState = $fields->herostate->value;
$heroLayout = $fields->herolayout->value;
$heroAltTitle = $fields->herotitle->value;
$titleColour = $fields->heroheadercolour->value;
$heroCopy = $fields->heroheadercopy->value;
$overlayColour = $fields->overlaycolour->value;
$overlayOpacity = $fields->overlayopacity->value;
$heroImage = article($articleID)->image->url;

if($heroState == true && !empty($heroImage)){
	page()->layout->header['image'] = $heroImage;
	page()->layout->header['heroState'] = true; 
	page()->layout->header['heroLayout'] = $heroLayout; 
	page()->layout->title = $pageTitle;
	if(!empty($heroAltTitle)) {
		page()->layout->titlealt = $heroAltTitle;
	}
	if(!empty($heroCopy)){
		page()->layout->herocopy = $heroCopy;
	}
	if(!empty($overlayColour)){
		page()->layout->overlaycolour = $overlayColour;
	}
	if(!empty($overlayOpacity)){
		page()->layout->overlayopacity = $overlayOpacity;	
	}
	if(!empty($titleColour)){
		page()->layout->titlecolor = $titleColour;
	}
	if(!empty($textColour)){
		page()->layout->textcolor = $textColour;
	}
}
```

### Layout
```php
$heroState = page()->layout->header->heroState;
$heroLayout = page()->layout->header->heroLayout;
$heroImage = page()->layout->header->image;
if($heroState == true){
	echo import('/partials/hero-header-' . $heroLayout . '.html.php', [
		'heroImage' => page()->layout->header->image,
		'title' => page()->layout->title,
		'titlealt' => page()->layout->titlealt,
		'herocopy' => page()->layout->herocopy,
		'overlaycolor' => page()->layout->overlaycolour,
		'overlayopacity' => page()->layout->overlayopacity,
		'parallax' => page()->layout->parallax,
		'titlecolor' => page()->layout->titlecolor,
		'textcolor' => page()->layout->textcolor
	]);
}
```

### Partial
```php
	$heroTitle   = $titlealt ?? $title ?? '';
	$heroCopy    = $heroCopy ?? '';
	$titlecolor  = $titlecolor  ?? '';
	$textcolor   = $textcolor ?? '';
	$heroImage   = $heroImage ?? '';
	
	$overlayClass = '';
	if(!empty($overlaycolour) || !empty($overlayopacity)) {
		$overlayClass = 'hasOverlay ' . $overlaycolour . ' ' . $overlayopacity;
	}

<section class="hero-header">
	<div class="hero-header__wrap hero-header__gradient py-12 px-4 bg-gradient-to-br from-yellow-300 to-cyan-400">
		<div class="flex flex-wrap items-center text-center lg:text-left -mx-2 w-full">
			<div class="lg:w-1/2 px-2 lg:pr-10 mt-10 lg:mt-0 order-1 lg:order-none">
				<div class="hero-header__wrap-inner ">
					<div class="<?= $textcolor; ?> font-sans text-right text-white py-8 pr-8">
						<h1 class="<?= $titlecolor; ?> text-7xl font-bold"><?= $heroTitle ?></h1>
						<? if($heroCopy):?>
						<div class="bg-primary py-8 pr-8 pl-20">
							<?= $herocopy; ?>
						</div>
						<? endif; ?>
					</div>
				</div>
			</div>
			<div class="lg:w-1/2 px-2"><img src="<?= $heroImage; ?>" alt=""></div>
		</div>
	</div>
</section>
```

I'm not sure how it could be better - put the code from the Page into an include? Layout doesn't matter as much because I virtually never see it. Same with the Partial. I guess I'm just after a sanity check before I start building out a ton of these things and wind up with an un-maintainable mess.

**Comment by johanjanssens** [comment]
I moved this to general, re-added the general section, just to keep the overview a bit since this isn't really a Q&A. Don't have a good answer yet. 

Need to think a bit how to do this best, i see what you are trying to do, this should be easier, a lot easier. I think custom field groups could help here, would allow you to have a `layout` group, fetch all the custom fields for that group and inject them into the layout in one go.  The all you need is the partial, which can get the data from the layout, no need for the layout step. Will get back on this one.

**Comment by kbrookes** [comment]
Thanks @johanjanssens - I wasn't sure if I was doing this in an optimal way or not.

**Comment by johanjanssens** [comment]
Changed the topic. What you are trying to do is a bit more specific than just passing variables in partials. You are trying to use custom fields to set layout properties that influence the page layout. Quite specific use-case already.

Worked this out for you, here is how you can now very easily do this:

### Grabbing the custom field values 

You can grab the field values as an array using the new `values()` method. This will flatten the array to a associative array, `name => value`. 

`<? $values = collection()->fields->values()) ?>`

or 

`<? $values = article('my-article')->fields->values()) ?>`

I also added a `group()`method to get only the fields for a specific group. You can for example create a `layout` group, then only retrieve the values for this group.

`<? $values = collection()->fields->group('layout')->values()) ?>`

### Injecting the values into the layout

I made `page()->layout` on config object so it's easier to work with it. You can now do:

`<? page()->layout->set([key], [value])); ?>`

### Putting it all together

You could create a layout with has an `options` property, could add defaults in the frontmatter. For example:

```md
---
options: 
    color: blue
    title: My Awesome Default Title
---
```

Here i defined two options, you would have two corresponding fields with the same names in Joomla.  Then anywhere in a single item page you could do:

`<? page()->layout->set('options', collection()->fields->group('layout')->values()); ?>`

or, by just getting the article fields directly

`<? page()->layout->set('options', article('my-article')->fields->group('layout')->values()); ?>`

Once you have done that you can access them as follows:

`<?= page()->layout->options->color ?>`
`<?= page()->layout->options->title ?>`

Remember that you do not need to pass layout properties to your partials, you can just use `page()->layout->options` in the partial, this does make the partial a bit more coupled to the layout, if you don't want that you can do.

`<?= import('/partials/hero-header.html', ['options' => page()->layout->options]); ?>`

Then in your partial just access them as  an object

`$options->color`

or if you want to make them a simple array you can unbox the options

`<?= import('/partials/hero-header.html', ['options' => unbox(page()->layout->options])); ?>`

Then in your partial just access them as an array

`$options['color']`

Changes are committed, but also require a small change in Pages itself, see: https://github.com/joomlatools/joomlatools-pages/commit/0c2f9bd166eca673d9a2c39856e92a76ee5500dc

**Reply by kbrookes** [reply]
That sounds amazing! I'll be giving that a try almost immediately.

Thanks!

**Reply by kbrookes** [reply]
I'm probably reading the PR wrong, or have missed something else, but changing just the single file I can see on that link just blanks the entire frontend.

**Reply by johanjanssens** [reply]
You forgot to copy the changes I made in the joomla extension. Sorry about that, but thought that was a given :)

**Reply by kbrookes** [reply]
Ah sorry man, I didn't realise there were changes there as well, I'm a bit of a Github repo newbie.

**Reply by kbrookes** [reply]
I know this is almost certainly me over-complicating things, but this bit caught my eye:

```<? page()->layout->set([key], [value])); ?>```

Which is saying to me that I can add key/value pairs into the layout from the page.

For example, although the image for my hero header example is not taken from custom fields (it's from the article image), it makes sense for it to be *grouped* with those options, because it forms part of that layout.

Is there a way of doing something like this:

```page()->layout->set('options', ([key],[value]));```

**Reply by kbrookes** [reply]
To clarify I don't **need** to group them together... just thought it might make sense to be able to set layout groups like that.

**Reply by johanjanssens** [reply]
Yes you can but the code is slightly different.  The `values()`method returns an array, you can just use the array to add a property before you inject it in the layout

```php
<? 
$values = collection()->fields->group('layout')->values() 
$values[property] = $[value]

<? page()->layout->set('options', $values);
?>
```

Another approach would be to set the property in the `options` object.

`<? page()->layout->options->set([key], [value])); ?>`

or 

`<? page()->layout->options->[key] = [value] ?>`

This would only work after the options property has been added otherwise you will get a notice error.  I have a few more ideas how to further improve this to make all this even more fluent. Working on those now.

**Reply by kbrookes** [reply]
That works brilliantly thanks - updated my Page to this:

page()->layout->set('options', article(1)->fields->group('Hero Headers')->values());
page()->layout->options->set('image', article($articleID)->image->url);

From there, was able to reference $options->image in the Partial without a problem.

**Reply by johanjanssens** [reply]
I have improved this a bit more for you to resolve the problem where a layout property doesn't exist yet, because the layout it was defined it wasn't loaded yet. 

Pages will now recursively merge all the layout properties before the page starts rendering, this means that even properties that have been defined in inherited layouts are available in the page itself. 

Also added a `layout()` template function to give you easy access to the layout properties. This function is advised to be used from now on as page()->layout will stop working in 0.20.

I provided more info as part of the following PR: https://github.com/joomlatools/joomlatools-pages/pull/540

I have a few further improvements to layouts planned by going to make those in the 0.20 release some of the changes will be breaking so don't want to do those now. You can find the ticket for this here: https://github.com/joomlatools/joomlatools-pages/issues/547

If you have further ideas or questions regarding layouts let me know!

**Comment by kbrookes** [comment]
This is all working much more nicely now thanks to @johanjanssens's rewrites and advice. I've got it down to something much simpler. On the Pages page, I have this:

```php
$article = article(1);

layout()->header['image'] = $article->image->url; // Set the article image to be the header image
layout()->set('options', $article->fields->group('Hero Headers')->values()); // Add the hero header field group to the layout group 'options'
```

This defines the fields, and adds the elements that are not in the Hero headers field group to the layout. It's greatly reduced from the original code.

The Layout itself is now vastly simpler also:

```php
$heroState = layout()->options->herostate; // Get hero state from options
$heroLayout = layout()->options->herolayout; // get hero layout from options
if($heroState == true){
    echo import('/partials/hero-header-' . $heroLayout . '.html.php', ['options' => layout()->options]);
}
```
It grabs the two variables needed to determine if a hero is needed, and what layout should be used. I'm not sure why $options->herostate doesn't work here, but accessing page()->layout is almost as simple.

And finally the Partial is not greatly changed, only the initial PHP setup has been modified:

```php
$title = layout()->title;
$titlealt = $options->herotitle;
$heroTitle   = $titlealt ?? $title ?? '';
$titlecolor = $options->heroheadercolour ?? '';	
$heroCopy    = $options->heroheadercopy ?? '';
$heroImage   = layout()->header->image ?? '';
$overlaycolour = $options->overlaycolour ?? '';
$overlayopacity = $options->overlayopacity ?? '';

if(!empty($overlaycolour) || !empty($overlayopacity)) {
	$overlayClass = 'hasOverlay ' . $overlaycolour . ' ' . $overlayopacity;
}
```

I hope this all helps someone else facing a similar challenge.

**Reply by johanjanssens** [reply]
I have edited your code here to reflect the changes made in https://github.com/joomlatools/joomlatools-pages/pull/540 The code will require Pages 0.19.6

**Reply by johanjanssens** [reply]
Just released 0.19.6 for you; https://github.com/joomlatools/joomlatools-pages/releases/tag/v0.19.6
